### PR TITLE
Expanded Bodies Data

### PR DESCRIPTION
**Pull request by AunSiro**
Additional bodies of the Solar System added to bodies.py.
Additional data of existing and new bodies added, including ROI.

k parameter data from: https://naif.jpl.nasa.gov/pub/naif/generic_kernels/spk/planets/a_old_versions/de421_announcement.pdf

Radius of the Sphere of influence formula from Wikipedia.

Mean radius, semi-mayor axis and eccentricity from Wikipedia, except for Saturn's semi-mayor axis, which is from http://articles.adsabs.harvard.edu/cgi-bin/nph-iarticle_query?1994A%26A...282..663S&amp;data_type=PDF_HIGH&amp;whole_paper=YES&amp;type=PRINTER&amp;filetype=.pdf

### CODE REVIEW

**Comment by Juanlu001** [comment]
Thanks a lot for this pull request Siro! This is probably a good starting point for https://github.com/poliastro/poliastro/issues/150. I have several comments:

* I'm not sure if the semimajor axis and the eccentricity belong here. While properties like the mass and the planet mean radius are clearly property of the planets themselves, the parameters of the orbit are _a consequence of that body being in orbit around a certain attractor_. For instance, you can retrieve the osculating orbit using `ephem.planet_ephem`, as described here http://poliastro.readthedocs.io/en/latest/user_guide.html#where-are-the-planets-computing-ephemerides. One could argue that having the **mean** properties of the orbit attached to the bodies might be practical, but still I'm not convinced from a code point of view. What do you think?
* In any case, if we are adding more properties I think they should be put inside the `__init__` method. Adding them dynamically after instantiation is possible but can confound IDEs and other introspection tools.
* Instead of hardcoding the formula inside `bodies.py` it would be nice to have a separate function to compute the spehere of influence radius. I think it's fine if we leave it inside `bodies`, we probably don't need to create a separate module for that.
* Notice that you changed the value of `k` for the Earth and three tests failed because of small numerical differences. At this point I think we need to choose an authoritative reference to fix these values and document it somewhere, because a) I don't know where did I copy mine from and b) I don't know where did you copy yours :sweat_smile:
* Regarding the last point, _the gravitational parametes depend on the ephemerides_: each one of them uses slightly different values to perform the computations. I already pointed it out here https://github.com/brandon-rhodes/python-jplephem/issues/9 Taking all of these variations would perhaps be too complex but at least we have to keep that in mind, and document it somewhere.
* We need tests for the new things we include, in particular the function that computes the radius of the sphere of influence but also the numerical properties.
* No, we are not including Pluto :)

**Comment by Juanlu001** [comment]
Fix: you linked your source for the gravitational parameters in the first comment. However, notice that DE421 is not our default ephemeris, but DE430:

https://github.com/poliastro/poliastro/blob/v0.6.0/src/poliastro/cli.py#L25

When we switch to Astropy to compute them (#131) the situation will be the same:

http://docs.astropy.org/en/stable/coordinates/solarsystem.html#solar-system-ephemerides

**Comment by AunSiro** [comment]
- I think that we should include the mean orbital parameters because they are very lightweight, good enought for most hobby applications and first approaches and don't force users to download medium-sized aphemeris files . If these files are not included with poliastro because they are too heavy, the users should not be forced to dowload them later in order to use the interesting items. Also, having a back-up will make the package independent from the user ability to connect to the internet and the state of the source files, which cann not be taken as granted (as happened to me a couple of days ago, when the pages 404'd because of the problems with the servers).
While you are right about the orbital parameters being a consecuence of the parent body gravity, I think that storing them inside each planet is the most clean and organiced  way.
-I agree that the way you are describing for storing and generating these new data are cleaner and more consistent, so I will change it and create a commit.
- I am very surprised that the change in the k value made the test fail, I only added a coulpe of decimals (!?)
- I don't really understand anything in the paragraph about how the gravitational parametes depend on the ephemerides. I don't understand either the conversation in the issue you linked T_T
- I will read how the other test are built and try to create a new one for the next commit!
- I think we sould include Pluto for a simple reason: it is very popular and the aditional work and complexity is very low. In order to be fair, I suggest adding the other four main dwarf planets: Ceres, Eris, Haumea and Makemake. I don't mind searching the data for them. If their Ephemeris is not included in the file, I thing the mean Keplerian elements should work fine.


I have a couple of questions:
- Should the body objects have a link to their gravitational parent body object? (may be useful in the future if data about moons and extrasolar sistem are added)
- Should the data be written in the bodys.py like now, or better read from a text file that could display all the data in a table form, also including the origin of the data? (may be more readable or useful if additional data columns are later considered)

**Comment by jzmiller1** [comment]
I've been creating bodies for stars with exoplanets.  While it was useful to have the Sun and Earth predefined as a new user I think that it might be better let everyone create their own bodies as needed.  Perhaps some tools for create bodies from common representations (CSV, some NASA formats) would be helpful.

Data for stars and exoplanets often comes in terms of Solar, Earth or Jupiter masses and radii.  It might be worth thinking about making it easier to define bodies in these terms as well. This is an example of how I'm handling it now...maybe there's a better way I'm overlooking:

```python
from astropy import units as u
from astropy import constants as c
from poliastro.bodies import Body

G = c.G.to((u.km**3)/(u.kg * u.s**2))
TRAPPIST1 = Body.from_parameters(k=0.08*c.M_sun * G,
                                 name='TRAPPIST',
                                 symbol=None,
                                 R=0.114 * c.R_sun.to(u.km))
```

Would a subclass of Body that has a solar mass/radius interface make sense in the poliastro library or would that be something that users should have in their own codebase?

**Comment by AunSiro** [comment]
I think that would be a really nice interface! We are creating the bodies at the moment through the `from_parameters` method, maybe another way to implement your interface would be to have another creation method in the body object, called `from_relative_values` for example. That way, bodies created either way would share the same class.

Either with a different class or as a method, that would work great together with my proposals here, letting the users insert bodies into their projects in three different ways: from the `poliastro.bodies.Body.from_parameters()`,  from the `poliastro.bodies.Body.from_relative_values` or just importing some common predefined bodies, like the ones I propose here, simply through `poliastro.bodies.Mars`

The ability to create them from CSV or NASA formats is also a great idea, you know of any standard form of distributing these data?

**Comment by jzmiller1** [comment]
I didn't really get what the class method was doing.  Are you thinking something more like this?

```python
class Body(object):
    """Class to represent a body of the Solar System.
    """

    def __init__(self, k, name=None, symbol=None, R=0 * u.km, **kwargs):
        """Constructor.
        Parameters
        ----------
        k : Quantity
            Standard gravitational parameter.
        name : str
            Name of the body.
        symbol : str
            Symbol for the body.
        R : Quantity
            Radius of the body.
        """
        self.k = k
        self.name = name
        self.symbol = symbol
        self.R = R
        self.__dict__.update(kwargs)

    @classmethod
    @u.quantity_input(k=u.km ** 3 / u.s ** 2, R=u.km)
    def from_parameters(cls, k, name, symbol, R):
        return cls(k, name, symbol, R)

    @classmethod
    def from_relative(cls, masses, name=None, symbol=None, R=None, body='sun'):
        bodies = {'sun': {'mass': c.M_sun, 'radius': c.R_sun.to(u.km)},
                  'earth': {'mass': c.M_earth, 'radius': c.R_earth.to(u.km)},
                  'jupiter': {'mass': c.M_jup, 'radius': c.R_jup.to(u.km)}
                  }
        k = masses * bodies[body]['mass'] * c.G.to((u.km ** 3) / (u.kg * u.s ** 2))
        R *= bodies[body]['radius']
        kwargs = {'body': body, 'bodies': bodies}
        return cls(k, name, symbol, R, **kwargs)

    def __str__(self):
        return u"{0} ({1})".format(self.name, self.symbol)

    def _repr_latex_(self):
        """Creates a LaTeX representation.
        Used by the IPython notebook.
        """
        return self.__str__()
```

A NASA format might be an ephemeris file from HORIZONS:

http://ssd.jpl.nasa.gov/horizons.cgi#top

http://ssd.jpl.nasa.gov/horizons.cgi#results

I know poliastro installs https://pypi.python.org/pypi/jplephem so this may be something that is already possible.  I'm still figuring things out here.

**Comment by Juanlu001** [comment]
Wow, lots of ideas going on here! I will answer to all tomorrow, but for the time being, let me just state that Astropy uses approximate orbital elements by default which do not need to download the JPL ephemeris

http://docs.astropy.org/en/stable/coordinates/solarsystem.html#solar-system-ephemerides

based on this algorithm

https://github.com/liberfa/erfa/blob/448a4ec/src/plan94.c

which, by the way, is similar in intent to what poliastro 0.1 offered:

https://github.com/poliastro/poliastro/blob/v0.1.2/poliastro/octave/uplanet_2013.m

**Comment by Juanlu001** [comment]
Answering @AunSiro first:

> I think that we should include the mean orbital parameters because they are very lightweight, good enought for most hobby applications and first approaches and don't force users to download medium-sized aphemeris files . If these files are not included with poliastro because they are too heavy, the users should not be forced to dowload them later in order to use the interesting items. Also, having a back-up will make the package independent from the user ability to connect to the internet and the state of the source files, which cann not be taken as granted (as happened to me a couple of days ago, when the pages 404'd because of the problems with the servers).

Does my previous comment address your concerns? When I implement #131, all computations will be feasible offline.

> While you are right about the orbital parameters being a consecuence of the parent body gravity, I think that storing them inside each planet is the most clean and organiced way.

I'm still not sure, can be convenient but it is also redundant. If we are adding the semimajor axis and eccentricity, I don't see why we would not add the other orbital parameters (except the anomaly). And if we are adding those, why not adding the mean equinoctial values too? Different people will suggest having different parameters there, and the `Body` class can end up being a catch-all object.

Instead, I believe retrieving the ephemerides and then turn that into a `State` object where we can get the desired values, although a bit longer, is easier to maintain in the long term.

> I am very surprised that the change in the k value made the test fail, I only added a coulpe of decimals (!?)

The perils of floating point I guess...

> I don't really understand anything in the paragraph about how the gravitational parametes depend on the ephemerides. I don't understand either the conversation in the issue you linked T_T

The JPL ephemerides fit some polynomials to a handful of observed quantities, and they also have to adjust the values of the gravitational parameter of each body. Every edition has different values, see for instance the ones used in DE421 and DE430:

| Ephemeris | Earth GM (km3/s2) |
|-------|-----|
| DE421 | 398600.436233 |
| DE430 | 398600.435436 |

* DE421 https://naif.jpl.nasa.gov/pub/naif/generic_kernels/spk/planets/a_old_versions/de421_announcement.pdf
* DE430 https://ilrs.cddis.eosdis.nasa.gov/docs/2014/196C.pdf

This is why we have to:

1. Choose which value we want to use, document it and adjust the tests, or
2. Change the values with the ephemerides that are chosen.

Option 2 is much more complicated, so if we are refining the values we use (which I agree it's a good idea) we should probably use those in DE430 and document this fact somewhere.

> I think we sould include Pluto for a simple reason: it is very popular and the aditional work and complexity is very low. In order to be fair, I suggest adding the other four main dwarf planets: Ceres, Eris, Haumea and Makemake. I don't mind searching the data for them. If their Ephemeris is not included in the file, I thing the mean Keplerian elements should work fine.

Fair enough, if we're including other dwarf planets I agree :)

> Should the body objects have a link to their gravitational parent body object? (may be useful in the future if data about moons and extrasolar sistem are added)

Good catch! Maybe a `parent` attribute, empty for the Sun?

> Should the data be written in the bodys.py like now, or better read from a text file that could display all the data in a table form, also including the origin of the data? (may be more readable or useful if additional data columns are later considered)

I'm not sure, maybe yes? We can consider having a CSV with those values, I would not oppose that idea if it does not make the code much more complicated.

> Is there any instructions available on how to use the ephemeris file? Which data does it contain?

I'm not sure what do you mean, can you rephrase?

**Comment by Juanlu001** [comment]
Thanks @jzmiller1 for jumping in! I think having the ability to create objects in terms of others is a good idea. Regarding your question:

> I didn't really get what the class method was doing.

It is only checking that the units are in place. The method was introduced in #140 because I observed that the unit conversion and checking are `extremely` slow, so I wanted to leave an "unsafe" method to create bodies without checking the units.

Maybe we can have something like this?

```
TRAPPIST1 = Body.from_relative(
    parent=Sun,
    k=0.08,  # Relative to the Sun
    name='TRAPPIST',
    symbol=None,
    R=0.114)  # Relative to the Sun

```

**Comment by Juanlu001** [comment]
Edit: I noticed that there can be a confusion in using the `parent` argument when creating a body, because `TRAPPIST1` is defined in terms of solar masses but _does not orbit around the Sun_. Another approach:

```python
TRAPPIST1 = Body.from_relative(
    reference=Sun,
    parent=None,
    k=0.08,  # Relative to the Sun
    name='TRAPPIST',
    symbol=None,
    R=0.114)  # Relative to the Sun

TRAPPIST1D = Body.from_relative(
    reference=Earth,
    parent=TRAPPIST1,  # Orbits TRAPPIST1
    # ...,
    R=0.772)  # Relative to the Earth
```

**Comment by jzmiller1** [comment]
The parent parameter isn't in the existing code and isn't in this pull request.  What is it meant to be used for?

**Comment by Juanlu001** [comment]
It isn't in the existing code, but I suggested its creation after this
discussion :) It's mean to hold the central body (Sun for the Earth, Earth
for the Moon).

**Comment by AunSiro** [comment]
I am near to create a new commit with some changes to this pull request. @jzmiller1 , Should I include your proposal for a function to create a body from relative values of another body, or would you rather create a new pull request with it yourself so it gets linked to you?

**Comment by jzmiller1** [comment]
I can put it in once your PR is merged in unless you prefer to do it.

**Comment by AunSiro** [comment]
I have commited a new version, the only left thing to do is add aditional test to cover the new code.

**Comment by AunSiro** [comment]
I wrote some test that I think that cover most of the new content! Expecting feedback for next iteration or merge at this point!

File: src/poliastro/bodies.py [hunk 1]
```diff
@@ -9,15 +9,22 @@
...
+au = 149597870.700 * u.km
+body_dict = {}
 
 class Body(object):
     """Class to represent a body of the Solar System.
 
     """
-    def __init__(self, k, name=None, symbol=None, R=0 * u.km):
+    def __init__(self, k, name=None, symbol=None, R=0 * u.km,
+                 parent=None, orbit = None):
```
Comment (Juanlu001) [inline src/poliastro/bodies.py]: I still dislike the idea of adding an orbit to the body itself. The mean elements available at a certain epoch are somewhat useful, but having approximate ephemerides through Astropy and ERFA it looks like duplicated data to me. I would still like the `Body` class to be as simple as possible, removing the orbit and the associated imports.

File: src/poliastro/bodies.py [hunk 2]
```diff
@@ -30,17 +37,24 @@ def __init__(self, k, name=None, symbol=None, R=0 * u.km):
...
+        orbit: Orbit
+            orbit object of the body
 
         """
         self.k = k
         self.name = name
         self.symbol = symbol
         self.R = R
+        self.parent = parent
+        self.orbit = orbit
```
Comment (Juanlu001) [inline src/poliastro/bodies.py]: See above.

File: src/poliastro/bodies.py [hunk 3]
```diff
@@ -30,17 +37,24 @@ def __init__(self, k, name=None, symbol=None, R=0 * u.km):
...
+        self.soi = None
```
Comment (Juanlu001) [inline src/poliastro/bodies.py]: Why defer the computation of the SOI? We can do that in `__init__`.

File: src/poliastro/bodies.py [hunk 4]
```diff
@@ -52,9 +66,68 @@ def _repr_latex_(self):
 
         """
         return self.__str__()
+    def calculate_soi(self):
```
Comment (Juanlu001) [inline src/poliastro/bodies.py]: This can be a function outside of the `Body` class:

```
def calculate_soi(body, a, parent=None):
    # Parameter `a` is mandatory now
    # `parent` can be None if `body` already has it
    pass
```

File: src/poliastro/bodies.py [hunk 5]
```diff
@@ -52,9 +66,68 @@ def _repr_latex_(self):
...
+        The body must have a parent body and an orbit defined'''
+        if self.parent == None:
+            warnings.warn('Unable to calculate SOI without valid parent body')
+        elif self.orbit == None:
+            warnings.warn('Unable to calculate SOI without valid orbit assigned')
+        else:
+            self.soi= self.orbit.a * (self.k / self.parent.k) ** (2/5)
+
+_data_path = os.path.join(os.path.dirname(__file__), "bodies_data.txt")
+with open(_data_path, 'r', encoding='utf-8') as _raw_data :
```
Comment (Juanlu001) [inline src/poliastro/bodies.py]: This should be a function.

Also, is there a chance we can use the `csv` module from the stdlib here, or even pandas?

https://docs.python.org/3/library/csv.html

File: src/poliastro/bodies.py [hunk 6]
```diff
@@ -52,9 +66,68 @@ def _repr_latex_(self):
...
+            _orbit = Orbit.from_classical(_parent, _a, _ecc, _inc, _raan, _argp, _nu)
+        body_dict[_name] = Body.from_parameters(_k, _name, _symbol, _R, _parent, _orbit)
+    
+
+for _body_name in body_dict:
+    _body = body_dict[_body_name]
+    if _body.parent != None and _body.orbit != None:
+        _body.calculate_soi()
+
+Sun = body_dict['Sun']
```
Comment (Juanlu001) [inline src/poliastro/bodies.py]: Regarding the r_SOI of the major bodies: it's still questionable which value for the semimajor axis to take, but on the other hand this is just an approximate value. Do you think we can hardcode the important ones in the data file and leave the rest blank? There is a table in Battin (imperial system though).

File: src/poliastro/tests/test_maneuver.py [hunk 7]
```diff
@@ -55,7 +55,7 @@ def test_hohmann_maneuver():
 
     man = Maneuver.hohmann(ss_i, Earth.R + alt_f)
 
-    assert_quantity_allclose(ss_i.apply_maneuver(man).ecc, 0 * u.one, atol=1e-15 * u.one)
+    assert_quantity_allclose(ss_i.apply_maneuver(man).ecc, 0 * u.one, atol=1e-14 * u.one)
```
Comment (Juanlu001) [inline src/poliastro/tests/test_maneuver.py]: This one is failing because of the change in parameters?

File: src/poliastro/twobody/tests/test_states.py [hunk 8]
```diff
@@ -110,7 +110,7 @@ def test_convert_from_coe_to_rv():
     assert_quantity_allclose(inc, expected_inc, rtol=1e-4)
     assert_quantity_allclose(raan, expected_raan, rtol=1e-4)
     assert_quantity_allclose(argp, expected_argp, rtol=1e-4)
-    assert_quantity_allclose(nu, expected_nu, rtol=1e-6)
+    assert_quantity_allclose(nu, expected_nu, rtol=1e-5)
```
Comment (Juanlu001) [inline src/poliastro/twobody/tests/test_states.py]: Same question as above

File: src/poliastro/tests/test_bodies.py [hunk 9]
```diff
@@ -33,6 +37,42 @@ def test_body_printing_has_name_and_symbol():
...
+    expected_soi = 1 * u.km
+    assert_quantity_allclose(_soi.to(u.km), expected_soi)
+    
+def test_generalized_Titius_Bode_Law():
+    _bodies = ['Mercury', 'Venus', 'Earth', 'Mars', 'Ceres',
+               'Jupiter', 'Saturn', 'Uranus', 'Neptune']
+    au = 149597870.700 * u.km 
+    expected_approx = au * 0.2075 * 1.7327 ** arange(1,10)
+    obtained = []
+    for _name in _bodies:
```
Comment (Juanlu001) [inline src/poliastro/tests/test_bodies.py]: I would advise keeping the cyclomatic complexity of the tests as unity and instead parametrize them. See for example:

https://github.com/poliastro/poliastro/blob/v0.6.0/src/poliastro/tests/test_iod.py

File: src/poliastro/bodies_data.txt [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+#
+#
+#       Stars or parentless Bodies:
+#
+Sun , ☉      , 132712440041, 695700 , None, None
+#
+#
+#       Planets of the Solar System 
+#
+Mercury, ☿      , 22031.78 , 2439.7 , Sun,  0.38709927, 0.20563593,  7.00497902, 252.25032350,  77.45779628,  48.33076593
```
Comment (Juanlu001) [inline src/poliastro/bodies_data.txt]: If following what I said above, we would rather exclude the orbit parameters from here and instead put the radius of the SOI. In the future we might be interested in adding J2 but that is out of scope for this PR.

File: src/poliastro/bodies.py (hunk 6 above)
Comment (AunSiro) [inline src/poliastro/bodies.py]: I don't like hardcoding the values or r_SOI for two reasons:
- It is not a physical measurable property, but an artificial radius that in the context of patched conics we decide that works well enough. It is a consequence of other values that we already have, not something you could measure in real life.
- If we calculate it trough a function, we can obtain it for any body with mass and semi-major axis, wether or not its value has been documented before. This means that we can use patched conics for mission planning for anything, from Trappist System to Jovian satellites exploration, or even fictional systems.

File: src/poliastro/tests/test_maneuver.py (hunk 7 above)
Comment (AunSiro) [inline src/poliastro/tests/test_maneuver.py]: Yes, the added precision for K parameter in DE430 made the test fail. Since the documentation of the test quotes a reference for the expected values, I thought that relaxing the tolerance was a better approach than changing the expected values.

File: src/poliastro/twobody/tests/test_states.py (hunk 8 above)
Comment (AunSiro) [inline src/poliastro/twobody/tests/test_states.py]: Same answer...

File: src/poliastro/bodies.py (hunk 5 above)
Comment (AunSiro) [inline src/poliastro/bodies.py]: Yes, I agree with having it built into a function. I will search how to use those libraries, but it will require to expand the dependencies of Poliastro, are you ok with it?

File: src/poliastro/tests/test_bodies.py (hunk 9 above)
Comment (AunSiro) [inline src/poliastro/tests/test_bodies.py]: Ok, I will try to reformulate with a parametric approach...

File: src/poliastro/bodies.py (hunk 5 above)
Comment (Juanlu001) [inline src/poliastro/bodies.py]: Yeah, it's okay. `csv` is in the `stdlib`, `pandas` would have to be installed but it's not a great deal I think. Try to go the easier route first.

File: src/poliastro/bodies.py (hunk 6 above)
Comment (Juanlu001) [inline src/poliastro/bodies.py]: Yes, I agree we will need a function to compute them anyway, for the reasons you explained. Maybe hardcoding them in some cases and computing them in others is an unnecessary inconsistency...

**Comment by Juanlu001** [comment]
By the way, the conflicts appeared because I just merged #155. We can ignore them until the last moment.

**Comment by Juanlu001** [comment]
We finally made a decision regarding the use of planetary constants in poliastro.

https://github.com/poliastro/poliastro/pull/197#issuecomment-321748009

So probably this PR will have to be re-purposed to think only what do we do with the radius of the sphere of influence. I suggest we have a conversation about this in Cáceres, @AunSiro :)

**Comment by Juanlu001** [comment]
Also @AunSiro, notice that all this began with an email I sent you July 1st, 2013...

**Comment by Juanlu001** [comment]
Since we [switched to Astropy to compute ephemerides](https://github.com/poliastro/poliastro/pull/155/commits/e353c4e46#diff-80029d89874aeee488c5a7c40d6e1c6f) (which include an approximate model that does not require external downloads) and merged [our latest modifications to the `bodies` module](https://github.com/poliastro/poliastro/blob/97e0a4f/src/poliastro/bodies.py) this PR has conflicts. I am closing it and suggesting an alternative course of action in #150.

Thanks again @AunSiro for your effort, this started many changes in poliastro that will be useful for many :pray: it was, perhaps, too ambitious :)
